core/src/main/java/org/apache/accumulo/core/metadata/ValidationUtil.java

```diff
@@ -62,9 +62,12 @@ public static void validateRFileName(String fileName) {
 
   public static void validateFileName(String fileName) {
     Objects.requireNonNull(fileName);
-    if (!fileName.matches("[\\dA-Za-z._-]+")) {
-      throw new IllegalArgumentException(
-          "Provided filename (" + fileName + ") contains invalid characters.");
+    for (int i = 0; i < fileName.length(); i++) {
+      final char ch = fileName.charAt(i);
+      if (!(Character.isLetterOrDigit(ch) || ch == '-' || ch == '.' || ch == '_')) {
+        throw new IllegalArgumentException(
+            "Provided filename (" + fileName + ") contains invalid characters.");
+      }
```

Comment:
This would separate out the predicate, and I think makes it a bit more readable instead of constantly calling "charAt" and worrying about off-by-one errors on the loop variable.

```suggestion
    IntPredicate validFileChar = ch -> Character.isLetterOrDigit(ch) || ch == '-' || ch == '.' || ch == '_';
    if (!fileName.codePoints().allMatch(validFileChar)) {
      throw new IllegalArgumentException(
          "Provided filename (" + fileName + ") contains invalid characters.");
    }
```

Reply:
I initially tested a similar version to the suggested code (and re-verified with the suggested code). In the benchmark tests I used a for-loop version ran in roughly half the time and selected based on performance. I can update to use the more readable version and/or analyze further if needed.

Reply:
I would be interested in the benchmark numbers with the final implementation.

Reply:
The following benchmark used the final implementation:
```
Benchmark                                                   (splitsRfile)  Mode  Cnt    Score    Error  Units
GarbageCollectorPerformanceIT.benchmarkReferences                   1,100  avgt    3    1.385 ±  0.070  ms/op
GarbageCollectorPerformanceIT.benchmarkReferences:rfile/op          1,100  avgt    3    0.014 ±  0.001  ms/op
GarbageCollectorPerformanceIT.benchmarkReferences                  10,100  avgt    3    6.865 ±  3.233  ms/op
GarbageCollectorPerformanceIT.benchmarkReferences:rfile/op         10,100  avgt    3    0.007 ±  0.003  ms/op
GarbageCollectorPerformanceIT.benchmarkReferences                 100,100  avgt    3   40.292 ± 14.152  ms/op
GarbageCollectorPerformanceIT.benchmarkReferences:rfile/op        100,100  avgt    3    0.004 ±  0.001  ms/op
GarbageCollectorPerformanceIT.benchmarkReferences                1000,100  avgt    3  245.171 ± 46.164  ms/op
GarbageCollectorPerformanceIT.benchmarkReferences:rfile/op       1000,100  avgt    3    0.002 ±  0.001  ms/op
```

Reply:
How do I run the benchmark myself?

Reply:
The link to the benchmark is below. The `gc-test-update` branch has not been merged yet, it has updates from #3499. It should be possible to run with `mvn verify`:
https://github.com/dtspence/accumulo-jmh-test/tree/gc-test-update

Reply:
So, for 1000 iterations we went from 233ms to 245ms with this change, but we down from 521ms using the HEAD of 2.1 branch, am I reading that right?

Reply:
I unsure what the `validateFileName` change is doing to the overall GC references benchmark. The last number has variance and looking at it further now. The average time observed across multiple benchmark executions based on A/B for `validateFileName`, seemed to show some difference (i.e. 233ms vs 245ms). The HEAD difference from 521ms to 233ms/245ms is accurate and had a larger impact from the `URI` and `validateFileName` regex removal. In context of the `validateFileName` I had used a separate benchmark for the method itself.

Just as minor note the `splitRfile` parameter is saying (for the larger number) 1K splits w/100 r-files per split. I am investigating the iterations question further.